### DISCUSSION TITLE
simple mode?

### DISCUSSION THREAD

**Discussion by dai-shi** [live-codes/livecodes#538] [Ideas]
Thanks to some improvements, personally, I'm fairly comfortable with the current experience. But, it's a little bit hard for me to recommend it to others. (Especially, compared to stackblitz.) It's just my personal feeling and I think there's nothing wrong in livecodes. Still, I thought I should open this discussion.

There seems two aspects to overcome.
1. less is more
2. look & feel

I'm really bad at those things, and can't help much. But maybe something like "simple mode" might do.
For instance, I'm pretty happy with the simple design and limited capability provided by https://code.esm.sh/.
<img width="1017" alt="image" src="https://github.com/live-codes/livecodes/assets/490574/4936dfab-1af1-4fe3-9068-000d2096d6b3">

**Comment by hatemhosny** [comment]
Thank you @dai-shi . I really appreciate your constructive feedback.

The overall app design does need improvement indeed. Acknowledging that this is not an area I'm good at, I have opened an issue [here](https://github.com/live-codes/livecodes/issues/539) asking for help.

I understand the request for a simple mode with limited features and less bloat.
Currently, there is a [lite mode](https://livecodes.io/docs/features/lite). Its main aim was to allow embedding a light-weight playground (e.g. for docs websites) with limited features, for basic demos without requiring large downloads. It has a very basic code editor with just syntax highlighting and no auto-complete. In addition in strips down most of the app features but retains the ability to compile any of the supported languages.

examples:
https://livecodes.io/?lite
https://livecodes.io/?lite&tsx
https://livecodes.io/?lite&template=react
https://livecodes.io/?lite&x=id/gu643572xh6

I have not revisited this for a while. Do you think using the monaco editor with this mode covers your use case?
What else can be improved?

I'm really grateful for you giving time to help improve LiveCodes. 
Thank you.

**Reply by dai-shi** [reply]
Oh, the lite mode looks pretty good! I didn't notice it in docs and the query parameter.
I especially like the quick loading. (So, technically lazy loading more features would be possible?)
What's missing in my workflow is:
- editor (and vim mode for me)
- save & share
- (very specific, but) custom import map visibility
- more DX (typescript, auto completion, console logs), but this would be the balance between lite and full modes.

**Reply by hatemhosny** [reply]
The lite bundle is indeed smaller than the full app. Lazy loading is already being heavily utilized even for the full app.

However, most of the performance gain in lite mode is by avoiding to load Monaco editor (download, parse and execute ~ 5mb of JavaScript code). 

A lighter-weight editor is CodeMirror6 (used by default on mobile, ~500kb). 
The editor can be changed from App menu -> Editor Settings. User settings are saved for the following sessions.
It currently has basic auto-complete, however, I plan to add a TypeScript LSP for it to have more powerful autocomplete and intellisense. 
It also supports vim mode. 
You may also choose the editor you want for embedded playgrounds using the SDK (https://livecodes.io/docs/configuration/configuration-object#editor).

Another option for reducing UI clutter is a "Zen mode" toggle, like the one on the left lower corner in StackBlitz. When enabled, most of the UI elements are hidden, keeping the editors, result page and console. This is something I would consider.

Anyway, I will think about these and let you know.
Please feel free to add more suggestions or feedback.
Much appreciated.

**Reply by dai-shi** [reply]
> by avoiding to load Monaco editor

I see.

> A lighter-weight editor is CodeMirror6

Sounds great. Could it do vim mode??

> User settings are saved for the following sessions.

Well, it's just me, but I use incognito window. That's why I like storing things in URL.

> "Zen mode"

That sounds interesting. Though, I still need "save & share".

**Reply by hatemhosny** [reply]
> Sounds great. Could it do vim mode??

Yes, it does support vim mode.
App menu -> Editor Settings 
or direct link: https://livecodes.io/?screen=editor-settings
 Choose CodeMirror and enable vim mode

> Well, it's just me, but I use incognito window. That's why I like storing things in URL.

All configuration options can be set in query params 
e.g. https://livecodes.io/?editor=codemirror&editorMode=vim

Configuration object docs: https://livecodes.io/docs/configuration/configuration-object/
Query params docs: https://livecodes.io/docs/configuration/query-params/

> > "Zen mode"
> 
> That sounds interesting. Though, I still need "save & share".

Zen mode will keep the share button. You can save by keyboard shortcut (Cmd/Ctrl + S), when the editor is focused. 
In addition, Zen mode will be a runtime toggle with no need to reload, so you can easily disable it and access all features when you need to. 
I plan to add keyboard shortcuts for most functionality and (may be) a command palette.

**Reply by dai-shi** [reply]
So, shamelessly, I prove that I didn't read the docs thoroughly. That might be a hint for necessity for online help (tooltips?). I don't know. It might go against the simple UI/UX.

**Reply by dai-shi** [reply]
https://livecodes.io/?template=react&editor=codemirror&editorMode=vim

Nice. I hope it won't load monaco editor in this case.

Now, I would like to create a custom template `jotai`.

(I'm getting used to full mode... So, not sure about simple mode.)

**Reply by dai-shi** [reply]
It feels like codemirror's vim mode is limited. So, monaco editor is probably unavoidable.

**Reply by hatemhosny** [reply]
It won't load monaco so long as the query param contains `editor=codemirror`

You can save custom template from app menu -> save as -> template. Then app menu -> new -> my templates. 
You can save 1 as a default template. It loads by default if you disable the Welcome screen, or can be loaded directly from the Welcome screen.

**Reply by hatemhosny** [reply]
> That might be a hint for necessity for online help (tooltips?)

I agree there are some features that are not obvious. 

Here are some ideas for later reference:

- I'll think about places for unobtrusive tooltips.
- Some quick tips on the loading screen.
- App tour 
- A series of short videos to show off the features that might not be obvious.

**Reply by hatemhosny** [reply]
> It feels like codemirror's vim mode is limited.

I'm using this:
https://github.com/replit/codemirror-vim

You may want to report the limitations there. This can be helpful to everybody.

**Reply by hatemhosny** [reply]
> Now, I would like to create a custom template `jotai`.

Also note, since you don't mind just storing settings in URL, that you may combine share url with other params to act as a custom template 

https://livecodes.io/?editorMode=vim&x=id/6257zmusg3e

**Reply by dai-shi** [reply]
yeah, I was thinking about it.

**Comment by dai-shi** [comment]
So, it feels like livecodes provides almost everything I need.
But, back to the original discussion, it's more is less.
I can configure it for myself, but can't let others to do so.
I will try to look for my best "queryParams", but still simple mode might be valuable.
- something very opinionated by the author
- something the first user can use without reading docs
- in that sense, please forget about vim mode, which is not for everyone.

**Reply by hatemhosny** [reply]
I agree. I will give it some thought. 
Thank you.

**Comment by hatemhosny** [comment]
In case you find this useful, this is the default config:
https://github.com/live-codes/livecodes/blob/develop/src/livecodes/config/default-config.ts

**Comment by hatemhosny** [comment]
updated this issue (https://github.com/live-codes/livecodes/issues/539) to ask for help improving visual design and provide mock for UI to facilitate style edits.
Best designs will be included in LiveCodes as official themes.

Also posted on twitter: https://x.com/hatem_hosny_/status/1773104699941462361?s=20

**Comment by hatemhosny** [comment]
hi @dai-shi 
[This PR](https://github.com/live-codes/livecodes/pull/550) adds 2 new modes related to this discussion: **focus and simple modes**.
I would appreciate if you can have a look and let me know what you think.

The overall app design is still yet to be addressed.

**Reply by dai-shi** [reply]
Thanks. I will see what will happen with Jotai contributors.

**Reply by hatemhosny** [reply]
If you have no major comments, I will merge this and prepare a release after a few checks.
This should be ready by tomorrow. 
I will let you know, in case you want to use a released version rather than the preview URL.
Thanks a lot for the helpful suggestions.

**Reply by dai-shi** [reply]
Sounds good.

**Comment by hatemhosny** [comment]
This is now [released](https://github.com/live-codes/livecodes/releases/tag/v27) as [v27](https://v27.livecodes.io/) and is the current version running on [livecodes.io](https://livecodes.io/)

In addition, the [SDK](https://livecodes.io/docs/sdk/) is also updated to [v0.5 0](https://www.npmjs.com/package/livecodes/v/0.5.0) to use new configuration options.

Thank you.

**Comment by hatemhosny** [comment]
The UI is redesigned into a cleaner, more modern, more consistent and more accessible design.
In addition, the theme can be customised with custom color (which would be specially useful for embedded playgrounds)
The UI now also supports internationalization (translations for 12 UI languages)

See: 
https://github.com/live-codes/livecodes/releases/tag/v36
https://blog.livecodes.io/livecodes-gets-a-fresh-look-and-goes-multilingual
